src/cell_group.hpp

```diff
@@ -21,8 +23,25 @@ class cell_group {
 
     virtual void reset() = 0;
     virtual void set_binning_policy(binning_kind policy, time_type bin_interval) = 0;
-    virtual void advance(time_type tfinal, time_type dt) = 0;
-    virtual void enqueue_events(const std::vector<postsynaptic_spike_event>& events) = 0;
+    virtual void advance(time_type tfinal, time_type dt, std::size_t epoch) = 0;
+
+    // Pass events to be delivered to targets in the cell group in a future epoch.
+    // events:
+    //    An unsorted vector of post-synaptic events is maintained for each gid
+    //    on the local domain. These event lists are stored in a vector, with one
+    //    entry for each gid. Event lists for a cell group are contiguous in the
+    //    vector, in same order that input gid were provided to the cell_group
+    //    constructor.
+    // tfinal:
+    //    The final time for the current integration epoch. This may be used
+    //    by the cell_group implementation to omptimise event queue wrangling.
+    // epoch:
+    //    The current integration epoch. Events in events are due for delivery
+    //    in epoch+1 and later.
+    virtual void enqueue_events(
+            util::subrange_view_type<std::vector<std::vector<postsynaptic_spike_event>>> events,
```

Comment:
The subrange_view_type must then give us a reference into the `post_synaptic_spike_event list[][]`

Reply:
So the subrange must include all (but may have more) events than events that occur from the beginning of `epoch` until it's end at `tfinal` that must be delivered in `epoch+1` (and|or?) later, correct?

Reply:
Yes. The events being passed in by this interface are for delivery in any epoch *after* `epoch`, not just the `epoch+1`.

Reply:
The challenge here is that both the `advance` and `enqueue_events` methods are aware of which epoch they are in, and when that epoch ends. However, we do not have information about when any future epochs will end, so we can only "bin" events into two categories
1. will be delivered in this epch
2. will be deilvered later

Reply:
You will see in the `mc_cell_group` implementation, we maintain two lists:
1. *current events*: contains all events that must be delivered in the current epoch, and potentially events to be delivered in a future epoch.
2. "future events": events to be delivered in a future epoch.

Events to be deilvered in a future epoch can exist in both queues.